### PR TITLE
Add :map-of schema

### PR DESCRIPTION
**Pull request by jeans11**
#38 

```clojure
(m/validate [:map-of string? int?] {"age" 18})
;; true

(m/validate [:map-of keyword? int?] {"age" 18})
;; false

### CODE REVIEW

File: src/malli/core.cljc [hunk 1]
```diff
@@ -275,6 +275,47 @@
...
+  (reify IntoSchema
+    (-name [_] :map-of)
+    (-into-schema [_ properties childs opts]
+      (when-not (and (seq childs) (= 2 (count childs)))
+        (fail! ::invalid-map-of))
+      (let [[key-schema value-schema :as schemas] (mapv #(schema % opts) childs)
+            key-valid? (-validator key-schema)
+            value-valid? (-validator value-schema)
+            validate (fn [m]
+                       (reduce
```
Comment (ikitommi) [inline src/malli/core.cljc]: `reduce-kv` should be used

Comment (ikitommi) [reply src/malli/core.cljc]: also guard that the value is a `map?`

File: src/malli/core.cljc [hunk 2]
```diff
@@ -275,6 +275,47 @@
...
+                  (conj acc {:path   path
+                             :in     in
+                             :schema this
+                             :value  m
+                             :type   ::invalid-type})
+                  (reduce
+                    (fn [acc [i [key value]]]
+                      (let [in (conj in key)
+                            distance-path (+ i distance)
+                            key-explainer' (key-explainer (into path [distance-path 0]))
```
Comment (ikitommi) [inline src/malli/core.cljc]: I don't think this is right, the `:path` should point to the schema in a form, so `(conj [path distance])` and `(conj [path (inc distance)])`?

File: src/malli/core.cljc [hunk 3]
```diff
@@ -275,6 +275,47 @@
...
+                  key-explainer (partial -explainer key-schema)
+                  value-explainer (partial -explainer value-schema)]
+              (fn explain [m in acc]
+                (if-not (map? m)
+                  (conj acc {:path   path
+                             :in     in
+                             :schema this
+                             :value  m
+                             :type   ::invalid-type})
+                  (reduce
```
Comment (ikitommi) [inline src/malli/core.cljc]: `reduce-kv` here too

**Comment by ikitommi** [comment]
Thanks for the PR! Was doing the same but will revert and use this instead. Few comments attached. I'll add a transformer after it's merged.

Also, both Prismatic & JSON Schema support maps with both names keys and predicate keys. This might be merged later with `:map`.

File: src/malli/core.cljc (hunk 1 above)
Comment (jeans11) [reply src/malli/core.cljc]: Are you talking about this? https://github.com/metosin/malli/pull/41/files#diff-80b352d39081ba397dc83fc6fafe7a47R294

Comment (ikitommi) [reply src/malli/core.cljc]: Oh, the `map?` was there already, my bad

File: src/malli/core.cljc [hunk 4]
```diff
@@ -275,6 +275,47 @@
...
+                (if-not (map? m)
+                  (conj acc {:path   path
+                             :in     in
+                             :schema this
+                             :value  m
+                             :type   ::invalid-type})
+                  (reduce-kv
+                    (fn [acc key value]
+                      (let [in (conj in key)
+                            key-explainer' (key-explainer (conj path distance))
```
Comment (ikitommi) [inline src/malli/core.cljc]: These actually only use data which is available already at line 296->, and outside of the returned function. Lines 298 & 299 can create the concrete explainers already.

**Comment by ikitommi** [comment]
Looking good, one small change request

**Comment by ikitommi** [comment]
Thanks!!

**Comment by ikitommi** [comment]
Merged the Schema transformations into master and added support for `-transformer` to `:map-of` too.

**Comment by ikitommi** [comment]
related #43

**Pull request merged**